components/client/src/main/java/com/hotels/styx/client/Transport.java

```diff
@@ -15,76 +15,41 @@
  * Encapsulates a single connection to remote server which we can use to send the messages.
  */
 class Transport {
-    private final Id appId;
-    private final CharSequence originIdHeaderName;
-
-    public Transport(Id appId, CharSequence originIdHeaderName) {
-        this.appId = requireNonNull(appId);
-        this.originIdHeaderName = requireNonNull(originIdHeaderName);
-    }
-
-    public HttpTransaction send(LiveHttpRequest request, Optional<ConnectionPool> origin, Id originId) {
-        Observable<Connection> connection = connection(request, origin);
+    public HttpTransaction send(LiveHttpRequest request, ConnectionPool connectionPool) {
+        Observable<Connection> connection = connectionPool.borrowConnection();
 
         return new HttpTransaction() {
             @Override
             public Observable<LiveHttpResponse> response() {
                 return connection.flatMap(connection -> {
-                    Observable<LiveHttpResponse> responseObservable = connection.write(request)
-                            .map(response -> addOriginId(originId, response));
+                    Observable<LiveHttpResponse> responseObservable = connection.write(request);
 
                     return ResponseEventListener.from(responseObservable)
-                            .whenCancelled(() -> closeIfConnected(origin, connection))
-                            .whenResponseError(cause -> closeIfConnected(origin, connection))
-                            .whenContentError(cause -> closeIfConnected(origin, connection))
-                            .whenCompleted(() -> returnIfConnected(origin, connection))
+                            .whenCancelled(() -> closeIfConnected(connectionPool, connection))
+                            .whenResponseError(cause -> closeIfConnected(connectionPool, connection))
+                            .whenContentError(cause -> closeIfConnected(connectionPool, connection))
+                            .whenCompleted(() -> returnIfConnected(connectionPool, connection))
                             .apply();
                 });
             }
 
-            private synchronized void closeIfConnected(Optional<ConnectionPool> connectionPool, Connection connection) {
-                if (connection != null && connectionPool.isPresent()) {
-                    connectionPool.get().closeConnection(connection);
-                }
+            private synchronized void closeIfConnected(ConnectionPool connectionPool, Connection connection) {
+                connectionPool.closeConnection(connection);
             }
 
-            private synchronized void returnIfConnected(Optional<ConnectionPool> connectionPool, Connection connection) {
-                if (connection != null && connectionPool.isPresent()) {
-                    connectionPool.get().returnConnection(connection);
-                }
+            private synchronized void returnIfConnected(ConnectionPool connectionPool, Connection connection) {
```

Comment:
Don't we ever need to synchronize on something different than the current object? In that case, do we actually need to set 'closeIfConnected' and/or 'returnIfConnected' as synchronized methods or should we aim to ensure they are idempotent?

Reply:
Hey David. There is no longer need to synchronise on anything. I realised myself the `synchronized` modifiers were left in the methods. They should have been removed already. Perhaps you have commented an outdated version?

Reply:
Thinking about it, perhaps I should remove these methods altogether.

Reply:
oh, sure, I see your commit now right before my comment.   You can remove them but I don't see too much of a difference between both options to be honest... I don't dislike the 'extra indirection' myself.